### ISSUE TITLE
ExCmd (most probably) supressing errors and causes everything to hang

### ISSUE THREAD

**Issue by Cinderella-Man** [akash-akya/ex_cmd#25]
Hello :wave: 

I debugging an issue when my program hangs without raising/throwing any errors. I got it down to as small example as possible to ease reproducing.

So here is a simple function that uses `Stream` inside the `try do .. rescue .. end` clause:

```
  def start() do
      try do
        Stream.map(1..3, fn x -> x end)
        |> Stream.map(fn _x ->
          DateTime.from_unix!(12345)
        end)
        |> Stream.run()
      rescue
        err -> Logger.error(Exception.format(:error, err, __STACKTRACE__))
      end

      IO.inspect("finished")
  end
```

As expected, `"finish"` is printed to the console(I'm running this in a new project using `mix run ./lib/buggy.ex`)

Now when I tweak the value passed to the `DateTime.from_unix!/1` to a string (it will cause an error):

```
  def start() do
      try do
        Stream.map(1..3, fn x -> x end)
        |> Stream.map(fn _x ->
          DateTime.from_unix!("12345")
        end)
        |> Stream.run()
      rescue
        err -> Logger.error(Exception.format(:error, err, __STACKTRACE__))
      end

      IO.inspect("finished")
  end
```

I'm getting the error *caught*(by the `try`) and logged to the console as well as `"finished"`.

Now, I want to use the `ExCmd` in place of the hardcoded range:

```
  def start() do
    try do
      ExCmd.stream!(["echo", "1\n2\n3\n"])
      |> Stream.map(fn _x ->
        DateTime.from_unix!("1223")
      end)
      |> Stream.run()
    rescue
      err -> Logger.error(Exception.format(:error, err, __STACKTRACE__))
    end

    IO.inspect("finished")
  end
```

This is where things get weird - I'm not getting anything logged. The code hangs without catching the error, and it's not printing `"finished"`. The only output I see is:

```
$ mix run ./lib/buggy.ex
Compiling 1 file (.ex)

12:23:47.680 [debug] Command started. os pid: 53759
```

Just to make sure - this will work fine if there's no error thrown/raised:

```
  def start() do
    try do
      ExCmd.stream!(["echo", "1\n2\n3\n"])
      |> Stream.map(fn _x ->
        DateTime.from_unix!(1223)
      end)
      |> Stream.run()
    rescue
      err -> Logger.error(Exception.format(:error, err, __STACKTRACE__))
    end

    IO.inspect("finished")
  end
```

`"finished"` is getting printed.

---

Can you reproduce this behaviour? Is this expected? Is my analysis correct?

I would really appreciate any feedback, as I've been scratching my head for a couple of days now :wink:

**Comment by akash-akya** [comment]
Hi @Cinderella-Man, thanks for raising the issue.

This might be a bug, I have to take a close look. But if you know the maximum time the program might take to exit, you can just set `exit_timeout` (ex: `exit_timeout: 1_000` for 1s)

**Comment by akash-akya** [comment]
I see, understood, i'll push a fix soon. It would be helpful if you share the exact program you are running. Because the issue seems to be related to how the program flush its output.

**Comment by akash-akya** [comment]
I re read your recent comment.
 the exit_timeout is for how long we should wait for command to exit **after** we write all input or when there is exception, not how long your overall pipeline going to take. So you can still use it

**Comment by Cinderella-Man** [comment]
I tried to run the code from that branch (I appreciate the snippet to load the branch from repo as a dependency :wink: )

It appears to be working great - try-catch is now catching the error :sunglasses: Awesome work!

**Comment by akash-akya** [comment]
Cool, I'll merge and merge and a cut a release in few days along with few other changes.

**Comment by akash-akya** [comment]
@Cinderella-Man released `v0.10.0` with the fix and few other changes 🎉
